File: fitness-registration/nodejs/tests/registrationService.test.js
const { register, unregister } = require('../services/registrationService');
const { sequelize, FitnessClass, Registration } = require('../models');

beforeAll(async () => {
  await sequelize.sync({ force: true });
  await FitnessClass.create({ type: 'Yoga', dayOfWeek: 'Monday', time: '08:00', maxCapacity: 2 });
});

afterEach(async () => {
  await Registration.destroy({ where: {} });
});

afterAll(async () => {
  await sequelize.close();
});

test('register on main list', async () => {
  const msg = await register('user1', 1);
  expect(msg).toBe('Registered on main list');
  const count = await Registration.count({ where: { classId: 1, status: 'MAIN' } });
  expect(count).toBe(1);
});

test('register on waiting list', async () => {
  await register('user1', 1);
  await register('user2', 1);
  const msg = await register('user3', 1);
  expect(msg).toBe('Registered on waiting list, position 1');
  const count = await Registration.count({ where: { classId: 1, status: 'WAITING' } });
  expect(count).toBe(1);
});

test('promotion from waiting list', async () => {
  await register('user1', 1);
  await register('user2', 1);
  await register('user3', 1);
  await register('user4', 1);

  expect(await Registration.count({ where: { classId: 1, status: 'MAIN' } })).toBe(2);
  expect(await Registration.count({ where: { classId: 1, status: 'WAITING' } })).toBe(2);

  const msg = await unregister('user1', 1);
  expect(msg).toBe('Unregistered from main list');

  const mainList = await Registration.findAll({ where: { classId: 1, status: 'MAIN' }, order: [['id', 'ASC']] });
  expect(mainList.length).toBe(2);
  expect(mainList[1].userId).toBe('user3');

  const waitingList = await Registration.findAll({ where: { classId: 1, status: 'WAITING' }, order: [['position', 'ASC']] });
  expect(waitingList.length).toBe(1);
  expect(waitingList[0].userId).toBe('user4');
  expect(waitingList[0].position).toBe(1);
});

test('unregister from waiting list', async () => {
  await register('user1', 1);
  await register('user2', 1);
  await register('user3', 1);

  const msg = await unregister('user3', 1);
  expect(msg).toBe('Unregistered from waiting list');

  const waitingList = await Registration.findAll({ where: { classId: 1, status: 'WAITING' }, order: [['position', 'ASC']] });
  expect(waitingList.length).toBe(0);
});

test('multiple promotions', async () => {
  await register('user1', 1);
  await register('user2', 1);
  await register('user3', 1);
  await register('user4', 1);

  await unregister('user1', 1);
  await unregister('user2', 1);

  const mainList = await Registration.findAll({ where: { classId: 1, status: 'MAIN' }, order: [['id', 'ASC']] });
  expect(mainList.length).toBe(2);
  expect(mainList[0].userId).toBe('user3');
  expect(mainList[1].userId).toBe('user4');

  const waiting = await Registration.count({ where: { classId: 1, status: 'WAITING' } });
  expect(waiting).toBe(0);
});

File: fitness-registration/java/src/test/java/com/fitness/RegistrationServiceTest.java
package com.fitness;

import com.fitness.model.FitnessClass;
import com.fitness.model.Registration;
import com.fitness.repository.FitnessClassRepository;
import com.fitness.repository.RegistrationRepository;
import com.fitness.service.RegistrationService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
class RegistrationServiceTest {

    @Autowired
    private RegistrationService registrationService;

    @Autowired
    private FitnessClassRepository classRepository;

    @Autowired
    private RegistrationRepository registrationRepository;

    private Long classId;

    @BeforeEach
    void setUp() {
        registrationRepository.deleteAll();
        classRepository.deleteAll();
        FitnessClass fc = classRepository.save(new FitnessClass("Yoga", "Monday", "08:00", 2));
        classId = fc.getId();
    }

    @Test
    void testRegisterMainList() {
        String result = registrationService.register("user1", classId);
        assertEquals("Registered on main list", result);
        long mainCount = registrationRepository.countByClassIdAndStatus(classId, "MAIN");
        assertEquals(1, mainCount);
    }

    @Test
    void testRegisterWaitingList() {
        registrationService.register("user1", classId);
        registrationService.register("user2", classId);
        String result = registrationService.register("user3", classId);
        assertEquals("Registered on waiting list, position 1", result);
        long waitingCount = registrationRepository.countByClassIdAndStatus(classId, "WAITING");
        assertEquals(1, waitingCount);
    }

    @Test
    void testPromotionFromWaitingList() {
        registrationService.register("user1", classId);
        registrationService.register("user2", classId);
        registrationService.register("user3", classId);
        registrationService.register("user4", classId);

        assertEquals(2, registrationRepository.countByClassIdAndStatus(classId, "MAIN"));
        assertEquals(2, registrationRepository.countByClassIdAndStatus(classId, "WAITING"));

        registrationService.unregister("user1", classId);

        List<Registration> mainList = registrationRepository.findByClassIdAndStatusOrderByPosition(classId, "MAIN");
        assertEquals(2, mainList.size());
        assertEquals("user3", mainList.get(1).getUserId());

        List<Registration> waitingList = registrationRepository.findByClassIdAndStatusOrderByPosition(classId, "WAITING");
        assertEquals(1, waitingList.size());
        assertEquals("user4", waitingList.get(0).getUserId());
        assertEquals(1, waitingList.get(0).getPosition());
    }

    @Test
    void testUnregisterFromWaitingList() {
        registrationService.register("user1", classId);
        registrationService.register("user2", classId);
        registrationService.register("user3", classId);
        registrationService.register("user4", classId);

        String result = registrationService.unregister("user4", classId);
        assertEquals("Unregistered from waiting list", result);

        List<Registration> waitingList = registrationRepository.findByClassIdAndStatusOrderByPosition(classId, "WAITING");
        assertEquals(1, waitingList.size());
        assertEquals("user3", waitingList.get(0).getUserId());
        assertEquals(1, waitingList.get(0).getPosition());
    }

    @Test
    void testMultiplePromotions() {
        registrationService.register("user1", classId);
        registrationService.register("user2", classId);
        registrationService.register("user3", classId);
        registrationService.register("user4", classId);

        registrationService.unregister("user1", classId);
        registrationService.unregister("user2", classId);

        List<Registration> mainList = registrationRepository.findByClassIdAndStatusOrderByPosition(classId, "MAIN");
        assertEquals(2, mainList.size());
        assertEquals("user3", mainList.get(0).getUserId());
        assertEquals("user4", mainList.get(1).getUserId());

        assertEquals(0, registrationRepository.countByClassIdAndStatus(classId, "WAITING"));
    }
}
